### PR TITLE
Add local file searching to the search command

### LINKED ISSUE #80: Browse local files through chat commands
**Is your feature request related to a problem? Please describe.**
Currently it is not possible to see what files are available through the filesystem.

**Describe the solution you'd like**
A chat command to list files or search them. Listing them could be similar to OS style commands(e.g. ls, cd).

**Describe alternatives you've considered**
Files could also be discovered through the web interface, though I think that should be a separate feature on its own.

### PR DESCRIPTION
**Pull request by Jokler**
The search command will now search local files first and if it can't
find any fall back to youtube-dl's "ytsearch".

**Summary**
This PR fixes/implements the following **bugs/features**

* Search command will search local files before using 'ytsearch'
* Logging will block instead of dropping logs if it has to

**Closing issues**

Resolves #80.

### CODE REVIEW

File: src/bot/music.rs [hunk 1]
```diff
@@ -363,10 +255,15 @@ impl MusicBot {
...
-                if let Err(e) = self
-                    .add_audio(format!("ytsearch:{}", query.join(" ")), invoker.name)
-                    .await
-                {
+                let uri = if let Some(path_str) = self.find_local_file(&query).await {
+                    // FIXME: Since we use a rust string non-utf8 file names
+                    // will not work as expected
+                    format!("file://{}", path_str)
+                } else {
+                    format!("ytsearch:{}", query.join(" "))
```
Comment (fkaa) [inline src/bot/music.rs]: Use enums instead

File: src/bot/music.rs [hunk 2]
```diff
@@ -411,6 +308,170 @@ impl MusicBot {
         Ok(())
     }
 
+    pub async fn add_audio(&mut self, uri: String, user: String) -> anyhow::Result<()> {
+        let metadata = if uri.starts_with(FILE_PREFIX) {
```
Comment (fkaa) [inline src/bot/music.rs]: can be skipped with enums

File: src/bot/music.rs [hunk 3]
```diff
@@ -411,6 +308,170 @@ impl MusicBot {
...
+
+        Ok(())
+    }
+
+    async fn find_local_file(&self, query: &Vec<String>) -> Option<String> {
+        if let Some(music_root) = &self.music_root {
+            let mut largest = (None, 0);
+
+            'outer:
+            for entry in WalkDir::new(music_root) {
```
Comment (fkaa) [inline src/bot/music.rs]: label on same line?

**Review by fkaa — CHANGES_REQUESTED** [review]

File: src/bot/music.rs (hunk 3 above)
Comment (Jokler) [reply src/bot/music.rs]: I ran `cargo fmt` again now. :)

**Review by fkaa — APPROVED** [review]

**Pull request merged**